fix: use documented synch APIs

rzhao271 opened the pull request:
Downstream PR: https://github.com/zeromq/libzmq/pull/4680

This PR would help Microsoft repositories indirectly consuming and rebuilding wepoll, such as [microsoft/zeromq-prebuilt](https://github.com/microsoft/zeromq-prebuilt), to get into compliance with internal policies by using APIs only documented on learn.microsoft.com.

~~However, this PR also brings in a dependency on [synchronization.lib](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress), which raises the minimum required Windows version to Windows 8.~~
This PR uses condition variables, bringing the minimum required Windows version to Windows Vista.

Older version of this PR that was for the dist branch: https://github.com/piscisaureus/wepoll/pull/36

rzhao271 committed:
fix: use documented synch APIs

rzhao271 committed:
Disable C4191 warnings

rzhao271 committed:
Add wait object

rzhao271 commented:
The tests pass locally, but I notice an issue with the wait object in the following scenario

1. Assume there are two void ptrs A and B where A != B.
2. A awaits, setting the global wait object to A.
3. B awaits, setting the global wait object to B.
4. A spuriously wakes up, but the global wait object is B, so the loop exits because A != B.

rzhao271 committed:
Use signal object instead of wait object

rzhao271 commented on src/reflock.c:
```diff
@@ -15,31 +15,30 @@
-  if (status != STATUS_SUCCESS)
+  BOOL status = TRUE;
+
+  if (reflock__signal_object == address) {
```
I added the if condition, because if the signal object is different from this address, then
- it already prevents early exits for this address, and
- it could be in the process of signalling another address

rzhao271 reviewed (commented):
@piscisaureus I have updated this PR and it is ready for review.

rzhao271 commented:
Hi @piscisaureus, could I get a review on this PR? 
Also let me know if I should make a PR on the dist branch again instead.

rzhao271 commented:
I rebuilt the change on a new device and noticed that test-reflock does not pass anymore. I'll investigate my changes further.

rzhao271 committed:
fix reflock test

rzhao271 commented on src/reflock.c:
```diff
@@ -11,35 +11,33 @@
-  if (status != STATUS_SUCCESS)
-    abort();
+static void reflock__signal_event(reflock_t* reflock) {
+  long state = InterlockedAdd(&reflock->state, REFLOCK__SIGNAL);
```
I'm now using a new signal bit to signal when a specific reflock has had signal called on it.

dhanalla commented:
Thanks @rzhao271 for creating this PR. 
As part of our security compliance initiative, we are also working to address undocumented APIs, as these are utilized within the OpenJDK repository.

rzhao271 committed:
Switch to condition variables

rzhao271 commented:
I have switched to using condition variables and added another bitflag, giving the following advantages:

- No need for synchronization.lib
- The minimum required OS version is Windows Vista or Windows Server 2008
- Using two condition variables per reflock and two bitflags makes it clearer that the signal and await functions actually wait for each other
- Spurious wakeups are handled for both signal and await functions with the while loops and bitflags
- Tests pass even after I increment TEST_ITERATIONS in test-reflock.c to 200
- No noticeable performance impact between this version and the keyed event one with TEST_ITERATIONS=200; both take around a minute each on my device.

AlanBateman commented:
Is there any possibility of getting the changes merged here? Ideally OpenJDK wouldn't have a fork with the changes.
